### PR TITLE
[NEW] Allow user's default preferences configuration

### LINKED ISSUE #3887: Default settings for My Account -> Preferences
The default settings for all the options under My Account -> Preferences are now currently hard-coded (overridable by ENV var at certain time).  

Due to the varied installation and deployment scenarios, we need a reliable (and hopefully user friendly) way to specify these defaults - and not assume our choice is the one that the end user/deployer wants for their site.

### PR DESCRIPTION
**Pull request by goiaba**
@RocketChat/core 

Closes  #3887

This improvement allows to set default values to be used in user's preferences. It creates a new section named "Default User Settings" into "Accounts" section where default configurations can be set.

### CODE REVIEW

**Comment by goiaba** [comment]
@RocketChatApp

It would be nice if someone could review this PR. :-)

**Comment by danieljhochman** [comment]
+1 Would also love to see this merged.

**Comment by rodrigok** [comment]
@goiaba This will change the values to the default ones just when the user save his profile.

Using that solution you will update the values and lost the state `undefined`, so if we change one setting on admin from, lets say, `false` to `true` there is no way to update the users to the new value cuz we can't know if the user has no value defined or if he set the value to `false` and don't want to change it to `true`.

The correct implementation for this is create a helper function to return the user's setting instead of accessing the `user.preferences.settings...`, the helper should receive the user object of id and the preference name/path to return, check the preference and return it if exists or return the setting from admin area if it does not exists.

What do you think @goiaba ?

**Comment by goiaba** [comment]
@rodrigok You mean we should replace every `users.settings.preferences.<a-key>` in the code by something like `getUserPreference(userId, preferenceKey [,fallbackValue])`?

The getUserPreference method would be responsible by choosing between the default value (when the user value is undefined) and the value the user set. If so, I like this approach.

**Comment by rodrigok** [comment]
@goiaba Exactly that :) thanks

File: tests/end-to-end/ui/11-admin.js [hunk 1]
```diff
@@ -704,5 +704,223 @@ describe('[Administration]', () => {
...
+			before(() => {
+				admin.settingsSearch.setValue('accounts');
+				admin.accountsLink.waitForVisible(5000);
+				admin.accountsLink.click();
+				admin.settingsSearch.setValue('');
+			});
+
+			describe('default user preferences', () => {
+				before(() => {
+					if (admin.accountsButtonCollapseDefaultUserPreferences.isVisible()) {
```
Comment (goiaba) [inline tests/end-to-end/ui/11-admin.js]: I'm wondering why this `if` is necessary here. Why are all the groups already expanded (and consequently the `expand` class does not exists in the html) when the test reach the Accounts section?

**Comment by goiaba** [comment]
@rodrigok I did some stuff here but need help to test accordingly. Too many changes...

File: packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js [hunk 2]
```diff
@@ -206,9 +206,9 @@ RocketChat.callbacks.add('afterSaveMessage', function(message, room, userId) {
 	subscriptions.forEach((s) => {
 		userIds.push(s.u._id);
 	});
-	const userSettings = {};
-	RocketChat.models.Users.findUsersByIds(userIds, { fields: { 'settings.preferences.audioNotifications': 1, 'settings.preferences.desktopNotifications': 1, 'settings.preferences.mobileNotifications': 1 } }).forEach((user) => {
-		userSettings[user._id] = user.settings;
+	const users = {};
+	RocketChat.models.Users.findUsersByIds(userIds, { fields: { 'settings.preferences': 1 } }).forEach((user) => {
```
Comment (rodrigok) [inline packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js]: This part is not good, we was getting only users that did change their configurations, not all users as this code does now.

File: packages/rocketchat-ui-account/client/accountPreferences.js [hunk 3]
```diff
@@ -56,52 +51,28 @@ Template.accountPreferences.helpers({
...
-	},
 	desktopNotificationDuration() {
-		const user = Meteor.user();
-		return user && user.settings && user.settings.preferences && user.settings.preferences.desktopNotificationDuration;
+		const userPref = RocketChat.getUserPreference(Meteor.user(), 'desktopNotificationDuration', 'undefined');
+		return userPref !== 'undefined' ? userPref : undefined;
 	},
 	defaultDesktopNotificationDuration() {
-		return RocketChat.settings.get('Desktop_Notifications_Duration');
+		return RocketChat.settings.get('Accounts_Default_User_Preferences_desktopNotificationDuration');
```
Comment (rodrigok) [inline packages/rocketchat-ui-account/client/accountPreferences.js]: Did you change `Desktop_Notifications_Duration` to `Accounts_Default_User_Preferences_desktopNotificationDuration` right? Is there a migration for that?

File: tests/end-to-end/ui/11-admin.js (hunk 1 above)
Comment (rodrigok) [reply tests/end-to-end/ui/11-admin.js]: I have no idea 😄

File: packages/rocketchat-ui-account/client/accountPreferences.js (hunk 3 above)
Comment (rodrigok) [reply packages/rocketchat-ui-account/client/accountPreferences.js]: Same for `Desktop_Notifications_Default_Alert` and `Mobile_Notifications_Default_Alert`

**Comment by rodrigok** [comment]
@goiaba It looks awesome, reviewing the code I had 2 main concerns:
1. Migration for certain values you changed
2. The sendNotificationsOnMessage can increase the process load cuz we are returning all users to iterate over now

File: packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js (hunk 2 above)
Comment (goiaba) [reply packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js]: @rodrigok Actually, as I didn't touch the `userIds` list and we are using a findByIds function, I think we are getting the same users, but now returning all their preferences. Or the `fields` argument generates something like a `where` clause?

File: packages/rocketchat-ui-account/client/accountPreferences.js (hunk 3 above)
Comment (goiaba) [reply packages/rocketchat-ui-account/client/accountPreferences.js]: Right, I forgot the migration script. :-( I Will work on it.

File: packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js (hunk 2 above)
Comment (rodrigok) [reply packages/rocketchat-lib/server/lib/sendNotificationsOnMessage.js]: 🤦‍♂️ You're right, was too much code to read 😄

**Pull request merged**